File: leetcode/POTD/2025-02-07/solution.py
class Solution:
    def queryResults(self, limit: int, queries: list[list[int]]) -> list[int]:
        result = []
        ball_to_color = {}  
        color_count = Counter()  

        for ball_id, new_color in queries:
            if ball_id in ball_to_color:
                previous_color = ball_to_color[ball_id]
                color_count[previous_color] -= 1
                if color_count[previous_color] == 0:
                    del color_count[previous_color]  

            ball_to_color[ball_id] = new_color
            color_count[new_color] += 1

            result.append(len(color_count))

        return result

File: leetcode/POTD/2025-02-07/solution.rs
use std::collections::HashMap;

impl Solution {
    pub fn query_results(_limit: i32, queries: Vec<Vec<i32>>) -> Vec<i32> {
        let mut result = Vec::new();
        let mut ball_to_color = HashMap::new();
        let mut color_count = HashMap::new();

        for query in queries.iter() {
            if query.len() != 2 {
                continue;
            }
            let ball_id = query[0];
            let new_color = query[1];

            if let Some(&previous_color) = ball_to_color.get(&ball_id) {
                if let Some(count) = color_count.get_mut(&previous_color) {
                    *count -= 1;
                    if *count == 0 {
                        color_count.remove(&previous_color);
                    }
                }
            }

            ball_to_color.insert(ball_id, new_color);
            *color_count.entry(new_color).or_insert(0) += 1;
            result.push(color_count.len() as i32);
        }

        result
    }
}
